### ISSUE TITLE
Raspberry Pi architecture is not detected correctly

### ISSUE THREAD

**Issue by mpsijm** [buildkite/agent#475]
Hi there,

I wanted to install a Buildkite agent on my Raspberry Pi. I thought I could just use the install script that you guys provide for Linux, but when I started the agent after that, it exited directly with the message `Illegal instruction`.
Since `uname -m` told me my architecture is `armv6l`, I tried manually downloading `buildkite-agent-linux-armhf-2.6.3.tar.gz` from the releases page on GitHub. Same thing happened: `Illegal instruction`.
Then I tried installing the `buildkite-agent-linux-arm-2.6.3.tar.gz` which, to my surprise, finally worked. My Buildkite agent is now happily "Waiting for work...".

TL;DR: Architecture `armv6l` should install the `arm` version instead of the `armhf` version.

**Comment by keithpitt** [comment]
@mpsijm 👋 I just had a sneaky look at fixing this. We have this list here:

https://github.com/buildkite/agent/blob/master/install.sh#L37

Would it be fair to change:

```
*armv6*)  ARCH="armhf" ;;
```

to...

```
*armv6*)  ARCH="arm" ;;
```

**Comment by mpsijm** [comment]
For my case, that will most likely fix the installation, but I'm not sure if you can guarantee that it will work for all ARM architectures.
After a small Google search, I found using `dpkg --print-architecture` is another alternative in order to which architecture someone is running. Maybe you can use that?
The output of this command is `armhf` on my Raspberry Pi and `amd64` on my Ubuntu laptop, so my guess is that you can directly set `ARCH=$(dpkg --print-architecture)` without a switch case (at least in the case of Linux).

**Comment by ryankurte** [comment]
Somewhat duplicated, the apt source on debian appears to install the wrong version as well. I got it working by adding an explicit architecture (as below), not sure what the correct fix is.

Broken: 
```
$ cat /etc/apt/sources.list.d/buildkite-agent.list
deb https://apt.buildkite.com/buildkite-agent stable main

$ apt search buildkite-agent
buildkite-agent/unknown,now 3.4.0-2484 armhf [installed]
...

$ buildkite-agent
Illegal instruction
```

Fix:
```
# Add arch to source
$ cat /etc/apt/sources.list.d/buildkite-agent.list
deb [arch=arm] https://apt.buildkite.com/buildkite-agent stable main

# Add arm to multiarch
$ sudo dpkg --add-architecture arm

# Update repos
$ sudo apt update

# Install with explicit arch
$ sudo apt install buildkite-agent:arm

$ buildkite-agent
Usage:

  buildkite-agent <command> [arguments...]
...
```

**Comment by lox** [comment]
I'd love to close this one out at some point. If someone who understands how arm's / linux's arch flags work comes across this, I'd love some advice!

**Comment by ryankurte** [comment]
this should fix the apparent problem, that said, the rpi _is armhf_ / does support hard float, so maybe there's something wrong with the build output too?

**Comment by lox** [comment]
We definitely are setting `CGO_ENABLED=0`: 

https://github.com/buildkite/agent/blob/3f8f6022ab76da5a3503161868435aa42a1889c4/scripts/build-binary.sh#L40

Perhaps the issue is that we are setting the wrong value for GOARM on

https://github.com/buildkite/agent/blob/3f8f6022ab76da5a3503161868435aa42a1889c4/scripts/build-binary.sh#L21

**Comment by lox** [comment]
Reading https://wiki.debian.org/RaspberryPi leaves me kind of at a loss in how to target the correct packages for different RPis.

**Comment by lox** [comment]
Do we just need to add much more specific arch's, like `armv5l`, `armv6l`, `armv7l` and  `armv5f`, `armv6f`, `armv7f`?

**Comment by ryankurte** [comment]
just got around to installing another agent, and confirmed this is now working on raspbian with agent version `v3.10.1` ^_^
